mlrun/datastore/datastore_profile.py

```diff
@@ -0,0 +1,143 @@
+            return None
+        return ast.literal_eval(self.value)["value"]
+
+
+class DatastoreProfile(pydantic.BaseModel):
+    type: str
+    name: str
+
+    @pydantic.validator("name")
+    def lower_case(cls, v):
+        return v.lower()
+
+
+class DatastoreProfileRedis(DatastoreProfile):
+    type: str = pydantic.Field("redis")
+    endpoint_url: str
+    username: typing.Optional[privateValue]
+    password: typing.Optional[privateValue]
+
+    @pydantic.validator("username", "password", pre=True)
+    def convert_to_private(cls, v):
+        return privateValue(value=v)
+
+    def is_secured(self):
+        return self.endpoint_url.startswith("rediss://")
+
+    def url_with_credentials(self):
+        parsed_url = urlparse(self.endpoint_url)
+        if self.username.get() and self.password.get():
+            netloc = (
+                f"{self.username.get()}:{self.password.get()}@{parsed_url.hostname}"
+            )
+        elif self.username.get():
+            netloc = f"{self.username.get()}@{parsed_url.hostname}"
+        else:
+            netloc = parsed_url.hostname
+
+        if parsed_url.port:
+            netloc += f":{parsed_url.port}"
+
+        new_parsed_url = ParseResult(
+            scheme=parsed_url.scheme,
+            netloc=netloc,
+            path=parsed_url.path,
+            params=parsed_url.params,
+            query=parsed_url.query,
+            fragment=parsed_url.fragment,
+        )
+        return urlunparse(new_parsed_url)
+
+
+class DatastoreProfile2Json(pydantic.BaseModel):
+    @staticmethod
+    def _to_json(attributes):
+        # First, base64 encode the values
+        encoded_dict = {
+            k: base64.b64encode(str(v).encode()).decode() for k, v in attributes.items()
+        }
+        # Then, return the dictionary as a JSON string with no spaces
+        return json.dumps(encoded_dict).replace(" ", "")
+
+    @staticmethod
+    def get_json_public(profile: DatastoreProfile) -> str:
+        return DatastoreProfile2Json._to_json(
+            {k: v for k, v in profile.dict().items() if not isinstance(v, dict)}
+        )
+
+    @staticmethod
+    def get_json_private(profile: DatastoreProfile) -> str:
+        return DatastoreProfile2Json._to_json(
+            {k: v for k, v in profile.dict().items() if isinstance(v, dict)}
+        )
+
+    @staticmethod
+    def create_from_json(public_json: str, private_json: str = "{}"):
+        attr1 = json.loads(public_json)
+        attr2 = json.loads(private_json)
+        attributes = {**attr1, **attr2}
+        decoded_dict = {
+            k: base64.b64decode(str(v).encode()).decode() for k, v in attributes.items()
+        }
+        if decoded_dict["type"] == "redis":
+            return DatastoreProfileRedis.parse_obj(decoded_dict)
+        else:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                "Datastore profie failed to create from json: Unexpected or missing type"
+            )
```

Comment:
Suggested improvement to error handling (and fix for typo):
```suggestion
        datastore_type = decoded_dict.get("type")
        if datastore_type == "redis":
            return DatastoreProfileRedis.parse_obj(decoded_dict)
        else:
            if datastore_type:
                reason = f"unexpected type '{decoded_dict['type']}'"
            else:
                reason = "missing type"
            raise mlrun.errors.MLRunInvalidArgumentError(
                f"Datastore profile failed to create from json due to {reason}"
            )
```